File: tot/memory/src/tot_memory/retrieval/decay.py
# SPDX-License-Identifier: GPL-2.0-or-later
"""Exponential time-decay for episodes, per design subspec §7.

Weight at age ``Δt`` (hours) with half-life ``h`` (hours):

    w(Δt) = 2 ** (-Δt / h)

→ ``w(0) = 1.0`` (fresh), ``w(h) = 0.5``, ``w(∞) → 0``.

Future timestamps (e.g. clock skew between the game server and the memory
sidecar) are clamped so they cannot produce ``w > 1``.

§7.2 specifies per-episode-type half-lives in seconds; this module exposes
them in hours because the public ``decay_weight`` signature takes
``half_life_hours`` (matches the plan-template's Task 15 contract). All
values are configurable via env vars at run time — see ``config.py``.
"""

import math
from datetime import datetime

# Per design subspec §7.2 — default half-lives in HOURS.
# Values converted from the §7.2 second-form for readability:
#   chat        48h    (172800s)
#   combat      72h    (259200s)
#   social      96h    (345600s)
#   quest      168h    (604800s — 7 days)
#   discovery  720h    (2592000s — 30 days)
#   goal       168h    (604800s — 7 days)
#   reflection 336h    (1209600s — 14 days)
#   observation 24h    (86400s)
HALF_LIVES_HOURS: dict[str, float] = {
    "chat":        48.0,
    "combat":      72.0,
    "social":      96.0,
    "quest":      168.0,
    "discovery":  720.0,
    "goal":       168.0,
    "reflection": 336.0,
    "observation": 24.0,
}

# Fallback when an unknown episode_type is supplied. The conservative choice
# is the shortest sensible default (chat — 48h); aging fast is safer than
# pinning irrelevant memories at the top of recall.
_FALLBACK_TYPE = "chat"


def half_life_for(episode_type: str) -> float:
    """Return the configured half-life in hours for ``episode_type``.

    Unknown types fall back to the ``chat`` half-life (48 h) so the
    caller always gets a usable value without raising.
    """
    return HALF_LIVES_HOURS.get(episode_type.lower(), HALF_LIVES_HOURS[_FALLBACK_TYPE])


def decay_weight(
    episode_time: datetime,
    now: datetime,
    half_life_hours: float,
) -> float:
    """Exponential decay multiplier in ``(0, 1]``.

    Args:
        episode_time: Event time of the episode (timezone-aware datetime).
        now:          Current wall-clock time.
        half_life_hours: Half-life in hours — the age at which decay = 0.5.
            Must be strictly positive.

    Returns:
        ``2 ** (-age_hours / half_life_hours)`` — clamped so that future
        timestamps (negative age) cannot exceed 1.0.

    Raises:
        ValueError: when ``half_life_hours <= 0``.
    """
    if half_life_hours <= 0:
        raise ValueError(f"half_life_hours must be positive, got {half_life_hours}")
    age_seconds = (now - episode_time).total_seconds()
    if age_seconds <= 0:
        return 1.0
    age_hours = age_seconds / 3600.0
    return math.pow(0.5, age_hours / half_life_hours)

File: tot/memory-rs/_archive-768/src/retrieval/decay.rs
// SPDX-License-Identifier: GPL-2.0-or-later
//! Exponential time-decay for episodes (design subspec §7 / §6.5).
//!
//! w(Δt) = 2^(-age_hours / half_life_hours)
//! age ≤ 0 (future timestamp) → 1.0   (clock-skew clamp)

/// Return the half-life in hours for a given episode type.
///
/// Unknown types fall back to `"chat"` (48 h) — the shortest sensible default
/// so stale unknown-type episodes age out rather than pinning at the top of
/// recall results.
pub fn half_life_for(episode_type: &str) -> f64 {
    match episode_type.to_lowercase().as_str() {
        "chat"        => 48.0,
        "combat"      => 72.0,
        "social"      => 96.0,
        "quest"       => 168.0,
        "discovery"   => 720.0,
        "goal"        => 168.0,
        "reflection"  => 336.0,
        "observation" => 24.0,
        _             => 48.0, // fallback = chat
    }
}

/// Exponential decay multiplier in (0, 1].
///
/// * `episode_time_ms` — event time as unix epoch milliseconds.
/// * `now_ms`          — current wall-clock as unix epoch milliseconds.
/// * `half_life_hours` — must be > 0; panics in debug, clamps to 1.0 in release.
///
/// Returns `2^(-age_hours / half_life_hours)` clamped so that future timestamps
/// (negative age) cannot produce a value > 1.0.
pub fn decay_weight(episode_time_ms: i64, now_ms: i64, half_life_hours: f64) -> f64 {
    if half_life_hours <= 0.0 {
        // Unreachable in production (all half-lives are positive constants).
        // Return 1.0 as a safe clamp rather than panicking in release builds.
        return 1.0;
    }
    let age_ms = now_ms - episode_time_ms;
    if age_ms <= 0 {
        return 1.0;
    }
    let age_hours = age_ms as f64 / 3_600_000.0;
    0.5_f64.powf(age_hours / half_life_hours)
}

#[cfg(test)]
mod tests {
    use super::*;

    // ── half_life_for ────────────────────────────────────────────────────────

    #[test]
    fn half_life_chat_is_48() {
        assert_eq!(half_life_for("chat"), 48.0);
    }

    #[test]
    fn half_life_combat_is_72() {
        assert_eq!(half_life_for("combat"), 72.0);
    }

    #[test]
    fn half_life_social_is_96() {
        assert_eq!(half_life_for("social"), 96.0);
    }

    #[test]
    fn half_life_quest_is_168() {
        assert_eq!(half_life_for("quest"), 168.0);
    }

    #[test]
    fn half_life_discovery_is_720() {
        assert_eq!(half_life_for("discovery"), 720.0);
    }

    #[test]
    fn half_life_goal_is_168() {
        assert_eq!(half_life_for("goal"), 168.0);
    }

    #[test]
    fn half_life_reflection_is_336() {
        assert_eq!(half_life_for("reflection"), 336.0);
    }

    #[test]
    fn half_life_observation_is_24() {
        assert_eq!(half_life_for("observation"), 24.0);
    }

    #[test]
    fn half_life_unknown_type_falls_back_to_48() {
        assert_eq!(half_life_for("not-a-real-type"), 48.0);
        assert_eq!(half_life_for(""), 48.0);
    }

    // ── decay_weight ─────────────────────────────────────────────────────────

    /// Fresh episode (age = 0) → weight 1.0 exactly.
    #[test]
    fn decay_weight_is_1_at_zero_age() {
        let now_ms: i64 = 1_748_000_000_000;
        assert_eq!(decay_weight(now_ms, now_ms, 24.0), 1.0);
    }

    /// Δt = one half-life → weight 0.5 (the defining property).
    #[test]
    fn decay_weight_is_half_at_one_half_life() {
        let now_ms: i64 = 1_748_000_000_000;
        let half_life_h = 24.0_f64;
        let ep_ms = now_ms - (half_life_h * 3_600_000.0) as i64;
        let w = decay_weight(ep_ms, now_ms, half_life_h);
        assert!((w - 0.5).abs() < 1e-9, "expected 0.5, got {w}");
    }

    /// Δt = 2 × half-life → weight 0.25.
    #[test]
    fn decay_weight_quarter_at_two_half_lives() {
        let now_ms: i64 = 1_748_000_000_000;
        let half_life_h = 24.0_f64;
        let ep_ms = now_ms - (2.0 * half_life_h * 3_600_000.0) as i64;
        let w = decay_weight(ep_ms, now_ms, half_life_h);
        assert!((w - 0.25).abs() < 1e-9, "expected 0.25, got {w}");
    }

    /// Future episode timestamp (clock skew) → weight clamped to 1.0.
    #[test]
    fn decay_weight_future_timestamp_clamps_to_1() {
        let now_ms: i64 = 1_748_000_000_000;
        let future_ms = now_ms + 3_600_000; // 1 hour in the future
        assert_eq!(decay_weight(future_ms, now_ms, 24.0), 1.0);
    }

    /// Verify each type's half-life round-trips through decay_weight correctly.
    #[test]
    fn decay_weight_per_type_half_life_round_trip() {
        let now_ms: i64 = 1_748_000_000_000;
        for (etype, expected_hl) in &[
            ("chat", 48.0_f64),
            ("combat", 72.0),
            ("social", 96.0),
            ("quest", 168.0),
            ("discovery", 720.0),
            ("goal", 168.0),
            ("reflection", 336.0),
            ("observation", 24.0),
        ] {
            let hl = half_life_for(etype);
            assert_eq!(hl, *expected_hl, "half_life_for({etype})");
            let ep_ms = now_ms - (hl * 3_600_000.0) as i64;
            let w = decay_weight(ep_ms, now_ms, hl);
            assert!(
                (w - 0.5).abs() < 1e-9,
                "type={etype}: expected 0.5 at one half-life, got {w}"
            );
        }
    }

    /// half_life_hours <= 0 → safe clamp (1.0), not a panic.
    #[test]
    fn decay_weight_zero_half_life_returns_one() {
        // Contract: unreachable in production; safe clamp rather than panic.
        let now_ms: i64 = 1_748_000_000_000;
        assert_eq!(decay_weight(now_ms - 1000, now_ms, 0.0), 1.0);
    }
}
